### DISCUSSION TITLE
Q About Usage of OBJ in Parameters

### DISCUSSION THREAD

**Discussion by tajmone** [AnssiR66/AlanStdLib#136] [Help and Support]
Ciao Annsi,

I wanted to ask you both a general question about the usage of `obj` in the library, as well as a specific verb case. I'm not sure I've fully grasped the criteria by which some verbs parameters are named as `obj` and others not.

For example (this is the specific case I mentioned), the verb `smell` names its parameter as `odour` (although in the comments it mentions `+ obj`):

```alan
-- ==============================================================


----- SMELL (+ obj)


-- ==============================================================


SYNTAX smell = smell (odour)
  WHERE odour ISA THING
```

Is it because `obj` in the library is reserved for physical objects which can be manipulated? and, in this case, one might smell an abstract thing (i.e. a smell `thing` created by the author to simulate a smell in the air, or somewhere in the location) as well as smelling a specific object (e.g. "smell apple")?

Somehow, I would have thought of the `smell` verb as primary targetting things, and would have expected `obj` as a parameter.

I'm asking you because as I translate the various verbs I notice that in the table of verbs not all of them have an associated `obj` although their parameter could potentially qualify as an object. So I wanted to ask you to explain me how the distinction between `obj` and non-`obj` parameters works — so that I might keep it consistent in my translation.

For example, since I'm translating parameters to Italian too, I could use the Italian word for "thing" ("cosa") without bumping into conflicts with Alan keywords (a luxury which in English can't be afforded). The same goes for "object", which I could freely translate to "oggetto" (I'm using `ogg` which is the shorthand for it, like `obj`).

> __NOTE__ — In the Italian library I've abolished the `smell0` restriction attribute, since both `smell0` and `smell` are the same verb I opted to use a common attribute for their restriction (if one is blocked so is the other).

**Comment by AnssiR66** [comment]
Hi Tristano!
the (+obj) in verb titles is just to indicate cases where there is an instance of a verb that is both transitive and intransitive, just like 'smell' and 'listen'.

>listen
You hear nothing....

>listen to man
You hear nothing...

This can be quite different from the actual parameter used. If it is confusing, we could list the verbs as smell (intransitive) and smell (transitive) instead. And I agree that 'odour' is a bit misleading, as smell takes normal things as objects anyway. It is mainly in congruence with some other codings (throw projectile, attack victim, etc) to increase the legibility of the code.  If you think there is reason to change this with 'smell', we can go ahead...

Best,
Anssi


________________________________

**Comment by tajmone** [comment]
I'm trying to understand more the last column of the Verbs Table (`OBJ`).

From what I gather, only verbs who have at least one parameter which is a `object` (ie, not of the `actor`, `thing`, `entity` or literal classes) get a checkmark in that column.

Why is it important to highlight this point? Is it related to reponse messages somehow?

Bare in mind that the Verbs Table in the Italian library also gets documented in the autogenrated HTML file:

- [`lib_verbi.html`](http://htmlpreview.github.io/?https://github.com/tajmone/Alan3-Italian/blob/master/alanlib_ita/lib_verbi.html#_tabella_comandi_di_partita)

So I wanted to add some explanation on the criteria for these columns, and why being able to tell which verbs have `object` parameters is important.

It's just that I feel that there is some important rationale (from the developing point of view) which I'm missing out here.

**Comment by AnssiR66** [comment]
Well, to be honest with you, I am not also sure why, but it was Thomas's wish to include it at that time and I forget the exact reason; without Thomas's suggestion I would probably  not have even considered having that column.
If I can find the old post about it, I will let you know...
Best,
Anssi
________________________________

**Comment by tajmone** [comment]
This might be worth digging it to improve the documentation for 2.2; maybe we should ask @thoni56 about the origin of the OBJ column in the verbs table.

It's not a crucial issue, but it would improve the overall understanding of the library for end users.

**Comment by thoni56** [comment]
Sorry, guys, I have no memory of that I even had such a conversation ,-) So I hope Anssi can dig up some old mails.

As the column is almost always marked where there is a parameter named 'obj' I can only guess that it has something to do with that name, and the only thing that comes to mind is the default syntaxes which uses that as its parameter name. But I have no idea why that would matter.

**Comment by tajmone** [comment]
Thanks Thomas,

> I can only guess that it has something to do with that name, and the only thing that comes to mind is the default syntaxes which uses that as its parameter name. But I have no idea why that would matter.

Yes, it's pointing out all the syntaxes which have a parameter (explicity) named `obj`. But I haven't so far worked out what are the benefits of having this in the verbs table. Probably it has to do with verbs overriding, when an author needs to know if the parameters are compatible (or maybe this was useful during the library development).

I was just hoping of providing some comments below the verbs table, illuminating the readers regarding its meaningfulness (having translated the verbs table in the Italian library, I've always been left wondering about that `OBJ` column purpose).

Maybe it will end up being an «heritage mistery» that carries on from one Library incarnation to the other, without anyone being wiser to its innermost secret, but diligently passing it on to the next generations as an act of legacy faith.

**Comment by AnssiR66** [comment]
Hi Thomas and Tristano,
I found this old mail from Sep 17 2010; Thomas, you sent me this one with the arity and obj columns added, after I sent you the verbs list with only the columns 'verb', 'synonym' and 'syntax' (below, I only quote the start of the list) :

-------------------------------------
>>Ok, here's an even better one with arity of the syntax and those having obj as parameter name. This way you can filter in various ways to get any interesting view:

VERB    SYNONYM SYNTAX  Arity   Obj
about   (+ help, info)  about   0

again   (+ g)   again   0

answer  (+ reply)       answer (str)    1

ask     (+ enquire, inquire, interrogate)       ask (act) about (topic) 2

attack  (+ beat, fight, hit, kick, punch)       attack (target) 1

attack_with
        attack (target) with (weapon)   2

bite    (+ chew)        bite (obj)      1       x
break   (+ destroy)     break (obj)     1       x
break_with
        break (obj) with (instr)        2       x
brief
        brief   0

burn
        burn (obj)      1       x
burn_with
        burn (obj) with (instr) 2       x
buy     (+ purchase)    buy (item)      1

catch
        catch (obj)     1       x
clean   (+ polish, wipe)        clean (obj)     1       x

------------------------------------------
(end of quoted mail)

But I guess we can get rid of the obj column if it doesn't serve any practical purpose?
-Anssi

**Comment by AnssiR66** [comment]
(of course, the original formulation in Thomas's  mail was much better but is lost in the above quote)

**Comment by tajmone** [comment]
Well done! You managed to recover this 9 years old email. Great.

So that's it, it boils down to being able to quickly filter which verbs take `obj` as a parameter?

I guess that `obj` being the most common (as well as the "natural") parameter this might be handy when there is the need to redefine verbs on classes, especially if you can pack similar-syntaxed verbs together.

I just can't come up with a decent paragraph explaining to the end user the helpfulness of that `OBJ` column — i.e. at least, not in a concise and technically correct way.

**Comment by tajmone** [comment]
@AnssiR66, I'm removing this Issue from the upcoming `v2.2`, leaving it for the next one that follows — these tables have been there for a long time, and to document/update them properly could take some time.

**Comment by AnssiR66** [comment]
Ok!

________________________________
